pre_commit/staged_files_only.py

```diff
@@ -74,7 +74,7 @@ def _unstaged_changes_cleared(patch_dir: str) -> Generator[None, None, None]:
                 # We failed to apply the patch, presumably due to fixes made
                 # by hooks.
                 # Roll back the changes made by hooks.
-                cmd_output_b('git', 'checkout', '--', '.', env=no_checkout_env)
+                cmd_output_b('git', 'restore', '--source', tree, '--worktree', '.', env=no_checkout_env)
```

Comment:
Hmmmm, part of me really wants a nicer message here in the case that we're on an older version of `git`, so that we can tell people how to fix themselves and prevent tools-help requests. Do you think adding an overridable message in the config for "you need a newer git" is overkill? (it very well might be lol)

I also think it's important to check that `restore` is available before we do anything, since otherwise we're in a data-loss scenario again, right? Eg, we'll only fail to to do the `restore` *exactly* at the time when we need it, and it's quite likely someone could have gone about all their other commit-doing-business before this without ever upgrading `git` itself.

Another option here is to do *another* `git write-tree` to save the current state of the index, do a `git checkout` which will update the index and the worktree, and then do a `git reset` to change the index *back* to its updated form. A bit ugly, but doesn't rely on a newer git version.

Either way I think a comment here about why we only want to change the worktree and leave the index in its current state is important (since I wrote out a whole comment about "why not just also reset the index back to what it was" before fully realizing the reason heh).

Reply:
> it's important to check that restore is available before we do anything

This is a _very_ good point.  All told, it's probably better to do the slightly ugly way that works on older git.  Switched to that and added a comment.